src/borg/archiver.py

```diff
@@ -1099,6 +1103,113 @@ def print_output(diff, path):
+        # ids = repository.list(limit=1, marker=None)
+        # cdata = repository.get(ids[0])
+        # key = key_factory(repository, cdata)
+        
+        if args.json:
+            def fixup_json(value):
+                if isinstance(value, dict):
+                    return { k.decode(): fixup_json(v) for k, v in value.items() }
+                
+                if isinstance(value, bytes):
+                    if len(value) < 65:
+                        return bin_to_hex(value)
+                    else:
+                        return f'{len(value)} bytes'
+
+                if isinstance(value, list):
+                    return [ fixup_json(v) for v in value ]
+
+                return value
+
+            def dump_item(id, data):
+                try:
+                    msg = msgpack.unpackb(data)
+                except UnpackException:
+                    msg = { b'data': data }
+
+                msg = fixup_json(msg)
+                msg['id'] = bin_to_hex(id)
+                json_print(msg)
+        else:
+            packer = msgpack.Packer()
+            def dump_item(id, data):
+                sys.stdout.buffer.write(packer.pack({
+                    b'id': id,
+                    b'data': data,
+                }))
+
+        marker = None
+        while True:
+            result = repository.scan(limit=LIST_SCAN_LIMIT, marker=marker)
+            if not result:
+                break
+            marker = result[-1]
+            for id in result:
+                cdata = repository.get(id)
+                give_id = id if id != Manifest.MANIFEST_ID else None
+                data = key.decrypt(give_id, cdata)
+                dump_item(id, data)
+
+        return self.exit_code
+
+    @with_repository(exclusive=True, compatibility=(Manifest.Operation.WRITE,))
+    def do_load(self, args, repository, manifest, key):
+        """load repository contents from stdin"""
+
+        # set up the key without depending on a manifest obj
+        # ids = repository.list(limit=1, marker=None)
+        # cdata = repository.get(ids[0])
+        # key = key_factory(repository, cdata)
+        
+        def add_archive(id, data):
+            try:
+                raw = msgpack.unpackb(data)
+            except msgpack.UnpackException:
+                return
+
+            if isinstance(raw, dict) and b'name' in data and b'version' in data:
+                archive = ArchiveItem(internal_dict=raw)
+                logger.info('Found archive %s', archive.name)
+                manifest.archives[archive.name] = (id, archive.time)
+
+        unpacker = msgpack.Unpacker(file_like=sys.stdin.buffer, object_hook=StableDict)
+        while True:
+            try:
+                item = unpacker.unpack()
+            except msgpack.OutOfData:
+                print('OUT OF DATA')
+                break
+            
+            if item[b'id'] == Manifest.MANIFEST_ID:
+                print('SKIP MANIFEST')
+                continue
```

Comment:
you are throwing away precious information (IDs of archives) here which you then need to determine again.

the problem is just that scan() very likely encounters the manifest **last** and you need it **first**. But as the manifest ID is known, that should be easy to special case.